src/libcsg/nblistgrid_3body.cc

```diff
@@ -0,0 +1,264 @@
+/* 
+ * Copyright 2009-2017 The VOTCA Development Team (http://www.votca.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <votca/csg/nblistgrid_3body.h>
+
+namespace votca { namespace csg {
+
+void NBListGrid_3Body::Generate(BeadList &list1, BeadList &list2, BeadList &list3, bool do_exclusions)
+{
+    BeadList::iterator iter;
+    _do_exclusions = do_exclusions;
+    if(list1.empty()) return;
+    if(list2.empty()) return;
+    if(list3.empty()) return;
+
+    //check if all bead lists "have" the same topology
+    assert(list1.getTopology() == list2.getTopology());
+    assert(list1.getTopology() == list3.getTopology());
+    assert(list2.getTopology() == list3.getTopology());
+    Topology *top = _top = list1.getTopology();
+
+    InitializeGrid(top->getBox());
+
+    // Add all beads of list1 to _beads1
+    for(iter = list1.begin(); iter != list1.end(); ++iter)
+        getCell((*iter)->getPos())._beads1.push_back(*iter);
+    
+    // Add all beads of list2 to _beads2
+    for(iter = list2.begin(); iter != list2.end(); ++iter)
+        getCell((*iter)->getPos())._beads2.push_back(*iter);
+    
+    // Add all beads of list2 to _beads3
+    for(iter = list3.begin(); iter != list3.end(); ++iter)
+        getCell((*iter)->getPos())._beads3.push_back(*iter);
+
+    //loop over beads of list 1 again to get the correlations
+    for(iter = list1.begin(); iter != list1.end(); ++iter) {
+        cell_t &cell = getCell((*iter)->getPos());
+        TestBead(cell, *iter);        
+    }
+    
+}
+    
+void NBListGrid_3Body::Generate(BeadList &list1, BeadList &list2, bool do_exclusions)
+{
+    BeadList::iterator iter;
+    _do_exclusions = do_exclusions;
+    if(list1.empty()) return;
+    if(list2.empty()) return;
+
```

Comment:
This is really similar to the code above can you not move it out into a single function?